### ISSUE TITLE
java.io.FileNotFoundException inside Bookkeeper post successful upgrade.

### ISSUE THREAD

**Issue by Richa-Srivastava** [pravega/pravega#4270]
**Once Pravega was upgraded from `0.6.0-2391.97363b3` to `0.6.0-2378.3ae8db3`, the upgrade was successful but Readers stalled from IO which was running in background**
```
INFO  [2019-10-11 07:59:12,840] io.pravega.longevity.utils.PerformanceUtils: Readers (4/6): events:949,769,714, events/sec:5,355, KB/sec:4106.326
INFO  [2019-10-11 07:59:12,840] io.pravega.longevity.utils.PerformanceUtils: Writers (3/3): events:949,786,677, events/sec:5,356, KB/sec:4107.0938
INFO  [2019-10-11 07:59:12,840] io.pravega.longevity.utils.PerformanceUtils: Readers (4/6): events:949,769,714, events/sec:5,356, KB/sec:4107.0205
INFO  [2019-10-11 07:59:42,839] io.pravega.longevity.utils.PerformanceUtils: Writers (3/3): events:949,937,540, events/sec:5,355, KB/sec:4106.3564
INFO  [2019-10-11 07:59:42,839] io.pravega.longevity.utils.PerformanceUtils: Readers (2/6): events:949,934,912, events/sec:5,355, KB/sec:4106.3457
INFO  [2019-10-11 07:59:42,839] io.pravega.longevity.utils.PerformanceUtils: Writers (3/3): events:949,937,540, events/sec:5,355, KB/sec:4106.3564
INFO  [2019-10-11 07:59:42,839] io.pravega.longevity.utils.PerformanceUtils: Readers (2/6): events:949,934,912, events/sec:5,355, KB/sec:4106.3457
INFO  [2019-10-11 08:00:12,838] io.pravega.longevity.utils.PerformanceUtils: Writers (3/3): events:950,103,001, events/sec:5,356, KB/sec:4106.377
INFO  [2019-10-11 08:00:12,838] io.pravega.longevity.utils.PerformanceUtils: Readers (1/6): events:950,100,107, events/sec:5,355, KB/sec:4106.365
```

Following is the upgrade status:
```
Status:
  Conditions:
    Last Transition Time:  2019-10-11T07:59:51Z
    Last Update Time:      2019-10-11T07:59:51Z
    Status:                False
    Type:                  Upgrading
    Last Transition Time:  2019-10-11T08:00:21Z
    Last Update Time:      2019-10-11T08:00:21Z
    Status:                True
    Type:                  PodsReady
    Last Transition Time:  2019-10-10T11:30:47Z
    Last Update Time:      2019-10-10T11:30:47Z
    Status:                False
    Type:                  Error
    Last Transition Time:  2019-10-10T11:30:47Z
    Last Update Time:      2019-10-10T11:30:47Z
    Status:                False
    Type:                  RollbackInProgress
  Current Replicas:        9
  Current Version:         0.6.0-2378.3ae8db3
```
I got following exception inside bookeeper:
```
2019-10-11 07:56:55,942 - ERROR - [BookieReadThreadPool-OrderedExecutor-1-0:ReadEntryProcessorV3@235] - IOException while reading entry: 24 from ledger 41137
java.io.FileNotFoundException: No file for log 265 for 41137 with location 2633104107922
        at org.apache.bookkeeper.bookie.EntryLogger.findFile(EntryLogger.java:1165)
        at org.apache.bookkeeper.bookie.EntryLogger.getChannelForLogId(EntryLogger.java:1100)
        at org.apache.bookkeeper.bookie.EntryLogger.internalReadEntry(EntryLogger.java:1002)
        at org.apache.bookkeeper.bookie.EntryLogger.readEntry(EntryLogger.java:1051)
        at org.apache.bookkeeper.bookie.InterleavedLedgerStorage.getEntry(InterleavedLedgerStorage.java:305)
        at org.apache.bookkeeper.bookie.SortedLedgerStorage.getEntry(SortedLedgerStorage.java:153)
        at org.apache.bookkeeper.bookie.LedgerDescriptorImpl.readEntry(LedgerDescriptorImpl.java:153)
        at org.apache.bookkeeper.bookie.Bookie.readEntry(Bookie.java:1305)
        at org.apache.bookkeeper.proto.ReadEntryProcessorV3.readEntry(ReadEntryProcessorV3.java:175)
        at org.apache.bookkeeper.proto.ReadEntryProcessorV3.readEntry(ReadEntryProcessorV3.java:155)
        at org.apache.bookkeeper.proto.ReadEntryProcessorV3.getReadResponse(ReadEntryProcessorV3.java:218)
        at org.apache.bookkeeper.proto.ReadEntryProcessorV3.executeOp(ReadEntryProcessorV3.java:264)
        at org.apache.bookkeeper.proto.ReadEntryProcessorV3.safeRun(ReadEntryProcessorV3.java:260)
        at org.apache.bookkeeper.common.util.SafeRunnable.run(SafeRunnable.java:36)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:748)
```
Also following exception inside SSS:
```
2019-10-11 07:58:35,489 70348 [BookKeeperClientWorker-OrderedExecutor-3-0] ERROR o.a.b.proto.PerChannelBookieClient - Read for failed on bookie pravega-bookie-2.pravega-bookie-headless.default.svc.cluster.local:3181 code EIO
2019-10-11 07:58:35,489 70348 [BookKeeperClientWorker-OrderedExecutor-3-0] ERROR o.a.bookkeeper.client.PendingReadOp - Read of ledger entry failed: L41135 E0-E39, Sent to [pravega-bookie-1.pravega-bookie-headless.default.svc.cluster.local:3181, pravega-bookie-2.pravega-bookie-headless.default.svc.cluster.local:3181, pravega-bookie-3.pravega-bookie-headless.default.svc.cluster.local:3181], Heard from [pravega-bookie-3.pravega-bookie-headless.default.svc.cluster.local:3181] : bitset = {0}. First unread entry is 1
2019-10-11 07:58:35,489 70348 [core-1] INFO  i.p.s.s.impl.rocksdb.RocksDBCache - RocksDBCache[Container_7]: Initialized.
2019-10-11 07:58:35,489 70348 [core-30] INFO  i.p.s.s.c.StreamSegmentContainerMetadata - SegmentContainer[5]: Exit RecoveryMode.
2019-10-11 07:58:35,489 70348 [core-30] INFO  i.p.s.server.CacheManager - CacheManager Unregistered ReadIndex[5-339078] (longevity/smallKey/692.#epoch.444).
2019-10-11 07:58:35,489 70348 [core-30] INFO  i.p.s.s.reading.ContainerReadIndex - ReadIndex[5]: Cleared.
2019-10-11 07:58:35,489 70348 [core-30] INFO  i.p.s.s.reading.ContainerReadIndex - ReadIndex[5] Exit RecoveryMode.
2019-10-11 07:58:35,490 70349 [core-4] INFO  i.p.s.s.r.StreamSegmentReadIndex - ReadIndex[5-339078]: Closed.
2019-10-11 07:58:35,491 70350 [BookKeeperClientWorker-OrderedExecutor-3-0] ERROR o.a.b.proto.PerChannelBookieClient - Read for failed on bookie pravega-bookie-1.pravega-bookie-headless.default.svc.cluster.local:3181 code EIO
2019-10-11 07:58:35,491 70350 [BookKeeperClientWorker-OrderedExecutor-3-0] ERROR o.a.b.proto.PerChannelBookieClient - Read for failed on bookie pravega-bookie-1.pravega-bookie-headless.default.svc.cluster.local:3181 code EIO
2019-10-11 07:58:35,491 70350 [core-30] ERROR i.p.s.server.logs.DurableLog - DurableLog[5] Recovery FAILED.
io.pravega.segmentstore.storage.DurableDataLogException: Error while reading from BookKeeper.
        at io.pravega.segmentstore.storage.impl.bookkeeper.LogReader.openNextLedger(LogReader.java:155)
        at io.pravega.segmentstore.storage.impl.bookkeeper.LogReader.getNext(LogReader.java:100)
        at io.pravega.segmentstore.storage.impl.bookkeeper.LogReader.getNext(LogReader.java:34)
        at io.pravega.segmentstore.server.logs.DataFrameInputStream.getNextFrame(DataFrameInputStream.java:310)
        at io.pravega.segmentstore.server.logs.DataFrameInputStream.getNextFrameEntry(DataFrameInputStream.java:290)
        at io.pravega.segmentstore.server.logs.DataFrameInputStream.fetchNextEntry(DataFrameInputStream.java:216)
        at io.pravega.segmentstore.server.logs.DataFrameInputStream.beginRecord(DataFrameInputStream.java:162)
        at io.pravega.segmentstore.server.logs.DataFrameReader.getNext(DataFrameReader.java:86)
        at io.pravega.segmentstore.server.logs.RecoveryProcessor.recoverAllOperations(RecoveryProcessor.java:159)
        at io.pravega.segmentstore.server.logs.RecoveryProcessor.performRecovery(RecoveryProcessor.java:93)
        at io.pravega.segmentstore.server.logs.DurableLog.performRecovery(DurableLog.java:215)
        at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1590)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.bookkeeper.client.BKException$BKReadException: Error while reading ledger
        at org.apache.bookkeeper.client.SyncCallbackUtils.finish(SyncCallbackUtils.java:83)
        at org.apache.bookkeeper.client.SyncCallbackUtils$SyncReadCallback.readComplete(SyncCallbackUtils.java:231)
        at org.apache.bookkeeper.client.LedgerHandle$6.onFailure(LedgerHandle.java:800)
        at org.apache.bookkeeper.common.concurrent.FutureEventListener.accept(FutureEventListener.java:38)
        at org.apache.bookkeeper.common.concurrent.FutureEventListener.accept(FutureEventListener.java:26)
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760)
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736)
        at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:442)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        ... 1 common frames omitted
```
Controller gave following exception:
```
2019-10-11 07:59:27,588 2929 [epollEventLoopGroup-2-1] WARN  i.p.client.netty.impl.RawClient - Closing connection with exception
io.pravega.shared.protocol.netty.ConnectionFailedException: WireCommands.WrongHost(type=WRONG_HOST, requestId=4294967296, segment=_system/_tables/scopes, correctHost=, serverStackTrace=)
        at io.pravega.client.netty.impl.RawClient$ResponseProcessor.process(RawClient.java:60)
        at io.pravega.client.netty.impl.FlowHandler.lambda$channelRead$2(FlowHandler.java:247)
        at java.util.Optional.ifPresent(Optional.java:159)
        at io.pravega.client.netty.impl.FlowHandler.channelRead(FlowHandler.java:245)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
        at io.pravega.shared.protocol.netty.ExceptionLoggingHandler.channelRead(ExceptionLoggingHandler.java:37)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1434)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:965)
        at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:799)
        at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:421)
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:321)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:897)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:748)
```
Following is the current Pravega Config:
```
root@osboxes:/usr/local/go/bin/src/github.com/pravega-longevity# kubectl get po -o jsonpath={...image}|tr -s [[:space:]] "\n"|sort|uniq
bookkeeper:0.6.0-2378.3ae8db3
pravega:0.6.0-2378.3ae8db3
openjdk:8
pravega/pravega-operator:0.4.2-rc2
pravega/zookeeper:0.2.3
pravega/zookeeper-operator:0.2.2
quay.io/external_storage/nfs-client-provisioner:v3.1.0-k8s1.11
pravega-grafana:latest
pravega-influxdb:latest
```
I am attaching metrics screen shot also:
![image](https://user-images.githubusercontent.com/48912640/66641537-972ad080-ec38-11e9-8325-0465f3ff9f1d.png)
![image](https://user-images.githubusercontent.com/48912640/66641563-9d20b180-ec38-11e9-9560-73cbae26e1a5.png)

Attaching required K8 logs on the corresponding Jira::Pravega-323.

**Comment by pbelgundi** [comment]
@Richa-Srivastava  we seem to be using different zookeeper and zookeeper-operator versions here:
```
pravega/zookeeper:0.2.3
pravega/zookeeper-operator:0.2.2
```
Why? Or was that a typo?

**Comment by pbelgundi** [comment]
@Richa-Srivastava Could you please share details about the longevity command you executed?
We would want to use the same when trying to reproduce the issue.

**Comment by Richa-Srivastava** [comment]
@pbelgundi Following is the command to run Longevity:
```
./tools/longevity_tests.py --image-repository=devops-repo.isus.emc.com:8116/nautilus/pravega-longevity --image-tag=26.5e8ec4b-0.6.0-2401.83b62b8 --controller-uri tcp://10.100.200.71:9090 --test=small-scale deploy
```

**Comment by pbelgundi** [comment]
From the uploaded logs I see that entire upgrade process had completed by `07:59` and bookie upgrade had completed by `7:55`.
I see this exception in bookie-1, bookie-2 and bookie-3 logs from `07:56:55` to `08:17:43`:
```
- ERROR - [BookieReadThreadPool-OrderedExecutor-1-0:ReadEntryProcessorV3@235] - IOException while reading entry: 36 from ledger 41137
java.io.FileNotFoundException: No file for log 25e for 41137 with location 2603352407124
        at org.apache.bookkeeper.bookie.EntryLogger.findFile(EntryLogger.java:1165)
        at org.apache.bookkeeper.bookie.EntryLogger.getChannelForLogId(EntryLogger.java:1100)
        at org.apache.bookkeeper.bookie.EntryLogger.internalReadEntry(EntryLogger.java:1002)
        at org.apache.bookkeeper.bookie.EntryLogger.readEntry(EntryLogger.java:1051)
        at org.apache.bookkeeper.bookie.InterleavedLedgerStorage.getEntry(InterleavedLedgerStorage.java:305)
        at org.apache.bookkeeper.bookie.SortedLedgerStorage.getEntry(SortedLedgerStorage.java:153)
        at org.apache.bookkeeper.bookie.LedgerDescriptorImpl.readEntry(LedgerDescriptorImpl.java:153)
        at org.apache.bookkeeper.bookie.Bookie.readEntry(Bookie.java:1305)
        at org.apache.bookkeeper.proto.ReadEntryProcessorV3.readEntry(ReadEntryProcessorV3.java:175)
        at org.apache.bookkeeper.proto.ReadEntryProcessorV3.readEntry(ReadEntryProcessorV3.java:155)
        at org.apache.bookkeeper.proto.ReadEntryProcessorV3.getReadResponse(ReadEntryProcessorV3.java:218)
        at org.apache.bookkeeper.proto.ReadEntryProcessorV3.executeOp(ReadEntryProcessorV3.java:264)
        at org.apache.bookkeeper.proto.ReadEntryProcessorV3.safeRun(ReadEntryProcessorV3.java:260)
        at org.apache.bookkeeper.common.util.SafeRunnable.run(SafeRunnable.java:36)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:748)
```
This could happen because of temporary inability to access the bookkeeper volumes during this period.
However I don't see any exceptions after `08:17:48`.
@Richa-Srivastava  could you please confirm if this was a transient issue or the readers stalled permanently post upgrade? Was there any data loss?

**Comment by pbelgundi** [comment]
@SrishT Could you please try to reproduce this issue with the correct `zookeeper/zookeeper-operator` versions and the above pravega versions.
It would help to get the events for bookie volumes and bookies during the period when we see the `FileNotFoundException`

**Comment by Richa-Srivastava** [comment]
@pbelgundi Readers stalled permanently post upgrade.

**Comment by SrishT** [comment]
@Richa-Srivastava could u tell me which client version did u trigger longevity with?

**Comment by SrishT** [comment]
I tried reproducing the issue with Pravega version `0.6.0-2378.3ae8db3`. IO started as expected when I ran longevity client version `26.5e8ec4b-0.6.0-2368.51b9974` on this cluster. However, even without upgrading the cluster, I observed a similar IO stall when I ran the cluster against longevity client version `26.5e8ec4b-0.6.0-2391.97363b3`.

Also confirmed with @shrids that the builds between `0.6.0-2362.7cfe76d` and `0.6.0-2396.026472c` are not cross compatible.

The PR which had a bug was: https://github.com/pravega/pravega/pull/4181
The PR which fixed this was : https://github.com/pravega/pravega/pull/4281

@Richa-Srivastava  could u retest this scenario with any build after `0.6.0-2396.026472c` and update the issue with your findings?

**Comment by pbelgundi** [comment]
This issue was observed again by @jkhalack when upgrading pravega from `0.6.1-2416` to `0.6.1-2417`.
PFA logs for the same.

**Comment by jkhalack** [comment]
I saw this updating from pravega `0.6.1-2416.1ac8306` to `0.6.1-2417.7b9ea59` (3 segmentstores, 1 controller, 3 zookeepers, 3 bookies)
Pravega-operator `0.4.3-rc1`
Zookeeper-operator `0.2.5-rc0`
Zookeeper `0.2.5-rc0`


No load on the system.
The controller is not able to restore the `_system` scope and further scope creation is failing:
```
2019-11-27 02:11:53,704 94509 [epollEventLoopGroup-2-2] WARN  i.p.c.s.e.ControllerEventProcessors - Error creating event processor scope _system with exception io.pravega.client.stream.impl.ControllerFailureException: Failed to create scope: _system
```
Otherwise the same kind of exceptions in the logs.

**Comment by pbelgundi** [comment]
@Richa-Srivastava Could you please update this issue with the pravega manifest used, if you still have it? Need to check on BK ensemblesize, write and ack quorum size

**Comment by Richa-Srivastava** [comment]
[pravega_manifest.txt](https://github.com/pravega/pravega/files/3903760/pravega_manifest.txt)
Attached is the yaml I used.

**Comment by eolivelli** [comment]
This is what we was looking for
bookkeeper.bkWriteQuorumSize: "3"
bookkeeper.bkAckQuorumSize: "3"

Just to double check:  Did the writer have errors during writes?

I think the problem is on the bookie:
- problem on disk (lost file?)
- bug in Bookkeeper

Here we have an old version of BK, and bookie internals changed a lot, it is difficult to track down to a bug

**Comment by RaulGracia** [comment]
Closed via #4420.
